File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
import sys
input = sys.stdin.readline

m,n = map(int,input().split())
target = [[0]* (m+1)]
for _ in range(m):
  temp = [0] + [int(x) for x in input().split()]
  target.append(temp)
  
sum_target = [[0]* (m+1) for _ in range(m+1)]

for i in range(1,m+1):
  for j in range(1,m+1):
    sum_target[i][j] = sum_target[i-1][j] + sum_target[i][j-1] - sum_target[i-1][j-1] + target[i][j]


answer = []
for _ in range(n):
  start_x, start_y,end_x, end_y = map(int,input().split())
  result = sum_target[end_x][end_y] - sum_target[start_x-1][end_y] - sum_target[end_x][start_y-1] + sum_target[start_x -1][start_y-1]
  answer.append(result)

print('\n'.join(map(str,answer)))

  

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer(br.readLine());
		int size = Integer.parseInt(st.nextToken()); // 배열 크기
		int t = Integer.parseInt(st.nextToken()); // 합 좌표 수
		
		int[][] map = new int[size+1][size+1];
		int[][] sumMap = new int[size+1][size+1];
		
		for(int i =1; i<= size ; i ++) {
			st = new StringTokenizer(br.readLine());
			map[i][0] = 0;
			for(int j = 1; j <= size; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		for(int i = 1 ; i <= size ; i ++) {
			for(int j = 1; j <= size; j++) {
				sumMap[i][j] = map[i][j] + sumMap[i-1][j] + sumMap[i][j-1] - sumMap[i-1][j-1]; 
			}
		}
		
		
		
//		for(int i = 0; i <=size; i++) {
//			System.out.println(Arrays.toString(sumMap[i]));;
//		}
		
		for(int i = 0; i < t;i++) {
			st = new StringTokenizer(br.readLine());
			int startX = Integer.parseInt(st.nextToken());
			int startY = Integer.parseInt(st.nextToken());
			int endX = Integer.parseInt(st.nextToken());
			int endY = Integer.parseInt(st.nextToken());
			int sum = sumMap[endX][endY] - sumMap[startX - 1][endY] - sumMap[endX][startY - 1] +sumMap[startX - 1][startY - 1];
			
			System.out.println(sum);
		}
	}
}
